### PR TITLE
[Quant][FX] Add Linear-(BN)-LeakyReLU fusion for ONEDNN backend

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Add Linear-(BN)-LeakyReLU fusion for FX quantization. It is enabled for ONEDNN backend only through `backend_config`.
```python
import torch
from torch.quantization.quantize_fx import prepare_fx, convert_fx
from torch.ao.quantization import get_default_qconfig_mapping
from torch.ao.quantization.backend_config import get_onednn_backend_config

qengine = 'onednn'
torch.backends.quantized.engine = qengine
qconfig_mapping = get_default_qconfig_mapping(qengine)
prepared_model = prepare_fx(model_fp32, qconfig_mapping, \
                            example_inputs=x, backend_config=get_onednn_backend_config())
quantized_model = convert_fx(prepared_model, backend_config=get_onednn_backend_config())
```

For FBGEMM and QNNPACK, FX won't fuse the pattern. If users set fbgemm/qnnpack as the backend while using onednn's `backend_config` to do quantization, an error is thrown when they run the quantized model.

Tests were added to make sure `linear - bn - leaky_relu` is not fused by default for fbgemm or qnnpack.

### CODE REVIEW

**Comment by vadimkantorov** [comment]
Related discussion: https://github.com/pytorch/vision/issues/4851#issuecomment-959745493, https://github.com/pytorch/pytorch/issues/26288, https://github.com/pytorch/pytorch/issues/23756, https://github.com/pytorch/vision/issues/4851

Given that BN + LeakyReLu+LeakyDropout can be made invertible, Conv+BN (or other normalization)+LeakyRelu+LeakyDropout can be fused or at least made into invertible modules and use no extra memory for storing activations (if a special invertible Sequential container is implemented that allows to explicitly prohibit hooks within this fused invertible container)

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 1]
```diff
@@ -269,6 +269,22 @@ at::Tensor& PackedLinearWeight::apply_relu_out(
   return apply_impl<true>(input, output_scale, output_zero_point, output);
 }
 
+at::Tensor PackedLinearWeight:: apply_leaky_relu(
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp:272]: I feel error out might make more sense because otherwise user would have silent failures, 

also for users who wants to use fbgemm and happens to have linear - leaky_relu pattern in the model, do they have to specify None for leaky_relu?

File: test/quantization/core/test_quantized_op.py [hunk 2]
```diff
@@ -3740,6 +3740,117 @@ def test_qlinear_unpack(self, W, use_channelwise):
...
+            batch_size *= 3  # Test the multi-dim input tensor
+        X_scale = 1.5
+        X_zp = 5
+        X_value_min = -128 if dtype == torch.qint8 else 0
+        X_value_max = 127 if dtype == torch.qint8 else 255
+        X_q0 = np.round(
+            np.random.rand(batch_size, input_channels) *
+            (X_value_max - X_value_min)
+            + X_value_min
+        ).astype(nptype)
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_op.py:3770]: can we use `_quantize` function? https://github.com/pytorch/pytorch/blob/master/test/quantization/core/test_quantized_op.py#L92

File: test/quantization/core/test_quantized_op.py [hunk 3]
```diff
@@ -3740,6 +3740,117 @@ def test_qlinear_unpack(self, W, use_channelwise):
...
+        if dtype == torch.qint8 or qengine_is_onednn():
+            W_zps = np.zeros(output_channels).astype(np.int)
+        else:
+            W_zps = np.round(np.random.rand(output_channels) * 100 - 50).astype(np.int)
+        # when using symmetric quantization
+        # special restriction for xnnpack fully connected op weight
+        # [-127, 127] instead of [-128, 127]
+        W_value_min = -127 if dtype == torch.qint8 else -128
+        W_value_max = 127
+        W_q0 = np.round(
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_op.py:3784]: similar for this one

File: test/quantization/core/test_quantized_op.py [hunk 4]
```diff
@@ -3740,6 +3740,117 @@ def test_qlinear_unpack(self, W, use_channelwise):
...
+                X_q0,
+                X_value_min,
+                X_value_max,
+                W_q0,
+                W_value_min,
+                W_value_max,
+            )
+        X = torch.from_numpy(_dequantize(
+            X_q0, X_scale, X_zp)).to(dtype=torch.float)
+        X_q = torch.quantize_per_tensor(
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_op.py:3809]: why do we dequantize and then quantize again here?

File: test/quantization/core/test_quantized_op.py [hunk 5]
```diff
@@ -3740,6 +3740,117 @@ def test_qlinear_unpack(self, W, use_channelwise):
...
 
+    """Tests the correctness of the quantized linear_leaky_relu op."""
+    @given(batch_size=st.integers(1, 4),
+           input_channels=st.integers(16, 32),
+           output_channels=st.integers(4, 8),
+           use_bias=st.booleans(),
+           use_multi_dim_input=st.booleans(),
+           use_channelwise=st.booleans())
+    @override_qengines
+    def test_qlinear_leaky_relu(self, batch_size, input_channels, output_channels, use_bias,
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_op.py:3751]: this is really long, can you add some comments to describe overall what we are testing here?

File: test/quantization/fx/test_quantize_fx.py [hunk 6]
```diff
@@ -4461,6 +4495,21 @@ def forward(self, x):
                 x = self.bn(x)
                 return x
 
+        class LinearBnLeakyReLUModel(torch.nn.Module):
+            def __init__(self, with_bn=True):
+                super(LinearBnLeakyReLUModel, self).__init__()
```
Comment (jerryzh168) [inline test/quantization/fx/test_quantize_fx.py:4500]: nit: `super().__init__()`

Comment (jerryzh168) [reply test/quantization/fx/test_quantize_fx.py:4500]: is this test module the same as the test module in test_fuse_linear_bn_leaky_relu_eval

File: torch/ao/quantization/backend_config/native.py [hunk 7]
```diff
@@ -260,6 +261,55 @@ def _get_linear_configs(dtype_configs):
         "root_module": torch.nn.Linear,
         "reference_quantized_module_for_root": nnqr.Linear,
     })
+
+    # (4) Linear + leaky_relu
```
Comment (jerryzh168) [inline torch/ao/quantization/backend_config/native.py:265]: wondering if we can dedup the code here as well, there must be some duplications with linear - relu pattern. I'm moving out this part of the code in https://github.com/pytorch/pytorch/pull/78407 maybe you can wait a little bit and rebase on that

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Xia-Weiwen) [reply aten/src/ATen/native/quantized/cpu/qlinear.cpp:272]: Below this line is the ref path for FBGEMM. I have added ref path for FBGEMM and QNNPACK. Users will go to ref path if they are using FBGEMM or QNNPACK.
Do you mean I should remove these and error out? I think that would be a problem since linear and leaky_relu will be fused regardless of the backend.

File: test/quantization/core/test_quantized_op.py (hunk 5 above)
Comment (Xia-Weiwen) [reply test/quantization/core/test_quantized_op.py:3751]: This test is based on `test_qlinear` just above: https://github.com/pytorch/pytorch/blob/299fbbccec68f9e8e0fa19a1066a8e179e9ab633/test/quantization/core/test_quantized_op.py#L3467
So, the structure and logic of the 2 UTs are the same, as well as usages of utils functions. This may also answer your questions below. Maybe we can extract some common parts of the two tests in the same test case `TestQuantizedLinear` https://github.com/pytorch/pytorch/blob/299fbbccec68f9e8e0fa19a1066a8e179e9ab633/test/quantization/core/test_quantized_op.py#L3457

File: test/quantization/fx/test_quantize_fx.py (hunk 6 above)
Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_fx.py:4500]: > nit: `super().__init__()`

This is to align with other sub-tests in the same test case: https://github.com/pytorch/pytorch/blob/299fbbccec68f9e8e0fa19a1066a8e179e9ab633/test/quantization/fx/test_quantize_fx.py#L4430

Comment (Xia-Weiwen) [reply test/quantization/fx/test_quantize_fx.py:4500]: > is this test module the same as the test module in test_fuse_linear_bn_leaky_relu_eval? should we use a single definition and add that to common_quantization.py? https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/common_quantization.py

Looks like they are doing similar things. Maybe I can remove `test_fuse_linear_bn_leaky_relu_eval`. Here I align with the sub-test of `LinearReLU`.
What do you mean by single definition in common_quantizaiton.py? Like this? https://github.com/pytorch/pytorch/blob/299fbbccec68f9e8e0fa19a1066a8e179e9ab633/torch/testing/_internal/common_quantization.py#L1323

File: torch/ao/quantization/backend_config/native.py (hunk 7 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/backend_config/native.py:265]: OK

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168. I have rebased this PR. Could you please review again at your convenience? Thanks a lot!

Notes:
- qlinear + q_leaky_relu will be fused for all backends.
- ONEDNN has a dedicated path while FBGEMM and QNNPACK only have a reference path (i.e., dequant - fp32 linear - fp32 leaky_relu - quant)
- Unit tests were updated per your comments.
  - A `LinearBnLeakyReluModel` model is defined in `common_quantization.py` for UT.
  - And `TestQuantizedLinear.test_qlinear_leaky_relu` is merged into `TestQuantizedLinear.test_qlinear` as a sub test.

**Comment by yanbing-j** [comment]
Hi @jerryzh168 , could you please help review this PR? Thank you .

**Comment by Xia-Weiwen** [comment]
This PR is currently pending on a solution on how to coordinate between the two quantization backends (FBGEMM, ONEDNN).

**Comment by kimishpatel** [comment]
Havent look in detail but I am worried that now suddenly leaky relu fused ops generated from quant workflows runs super slow. So I would like quant workflow to produce leaky-relu fused linear only when this is done for onednn/fbgemm backend. And not fuse them when qnnpack is the backend being used for quantizaiton.

**Comment by vkuzo** [comment]
> Havent look in detail but I am worried that now suddenly leaky relu fused ops generated from quant workflows runs super slow. So I would like quant workflow to produce leaky-relu fused linear only when this is done for onednn/fbgemm backend. And not fuse them when qnnpack is the backend being used for quantizaiton.

Agreed, this should be off by default for qnnpack unless there is a fast kernel.  IMO it could be cleaner to split this PR into something like
1. a kernel PR which adds fused linear_leaky_relu and a test that the kernel is numerically correct
2. a workflow PR which enables this fusion for fbgemm/onednn, but not for qnnpack (this may be blocked on the fusion API properly supporting backend differences, @jerryzh168  is there any work our team needs to do to enable this)

**Comment by Xia-Weiwen** [comment]
Hi, @kimishpatel, @vkuzo. Thanks for the comments. Since this PR is not updated for a while, the description was not correct. I have synced with @jerryzh168 and now the fusion is for onednn backend only. FX won't fuse linear + leaky_relu for fbgemm or qnnpack. If users manually call the fused op while using fbgemm/qnnpack, an error is thrown. How does that sound to you?

**Comment by vkuzo** [comment]
> Hi, @kimishpatel, @vkuzo. Thanks for the comments. Since this PR is not updated for a while, the description was not correct. I have synced with @jerryzh168 and now the fusion is for onednn backend only. FX won't fuse linear + leaky_relu for fbgemm or qnnpack. If users manually call the fused op while using fbgemm/qnnpack, an error is thrown. How does that sound to you?

Sounds reasonable, can we add a test that linear + leaky_relu is not fused by default for fbgemm and qnnpack?

**Comment by Xia-Weiwen** [comment]
> > Hi, @kimishpatel, @vkuzo. Thanks for the comments. Since this PR is not updated for a while, the description was not correct. I have synced with @jerryzh168 and now the fusion is for onednn backend only. FX won't fuse linear + leaky_relu for fbgemm or qnnpack. If users manually call the fused op while using fbgemm/qnnpack, an error is thrown. How does that sound to you?
> 
> Sounds reasonable, can we add a test that linear + leaky_relu is not fused by default for fbgemm and qnnpack?

Added

**Comment by Xia-Weiwen** [comment]
Hi @jerryzh168 Please review. Thanks.

File: torch/ao/quantization/backend_config/onednn.py [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+    reverse_sequential_wrapper2,
+    reverse3,
+)
+
+
+# ===================
+# |  DTYPE CONFIGS  |
+# ===================
+
+# Based on fbgemm's configs
```
Comment (jgong5) [inline torch/ao/quantization/backend_config/onednn.py:34]: Do we have to copy all the fbgemm configs here? onednn doesn't support all these configs, e.g., it doesn't support fp16 and quint4x2. How about only adding those onednn supports?

File: torch/ao/quantization/backend_config/onednn.py [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+        .set_observation_type(observation_type)  # noqa: E131
+        .set_dtype_configs(linear_dtype_configs))
+# linear leaky_relu, functional linear + functional leaky_relu
+linear_configs.append(
+    BackendPatternConfig((F.leaky_relu, F.linear))
+        .set_observation_type(observation_type)  # noqa: E131
+        .set_dtype_configs(linear_dtype_configs))
+
+'''
+# (2) Linear + tanh
```
Comment (jgong5) [inline torch/ao/quantization/backend_config/onednn.py:194]: Add it in a separate PR?

File: aten/src/ATen/native/quantized/cpu/qlinear.cpp [hunk 10]
```diff
@@ -621,9 +621,10 @@ at::Tensor PackedLinearWeightsQnnp::apply_relu(
 #endif // USE_PYTORCH_QNNPACK
 
 #if AT_MKLDNN_ENABLED()
-template <bool ReluFused>
+template <PostOps post_op>
 at::Tensor PackedLinearWeightsOnednn::apply_impl(
     at::Tensor input,
+    double negative_slope,
```
Comment (jgong5) [inline aten/src/ATen/native/quantized/cpu/qlinear.cpp:627]: Suggest not to add the post-op specific parameter into the args of a general function. Consider to do something like below:
https://github.com/pytorch/pytorch/pull/86583/files#diff-c24c006375f92e12dd5b062f143b62d41f347477b0efb7c25e3824ac41f0ad7eR184

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/ao/quantization/backend_config/onednn.py (hunk 8 above)
Comment (Xia-Weiwen) [reply torch/ao/quantization/backend_config/onednn.py:34]: OK I will fix these later.

File: test/quantization/core/test_quantized_op.py [hunk 11]
```diff
@@ -3814,6 +3814,111 @@ def test_qlinear_unpack(self, W, use_channelwise):
             np.testing.assert_equal(
                 W_q.q_zero_point(), W_q_origin.q_zero_point())
 
+    @given(batch_size=st.integers(1, 4),
+           input_channels=st.integers(16, 32),
+           output_channels=st.integers(4, 8),
+           use_bias=st.booleans(),
+           use_multi_dim_input=st.booleans(),
+           use_channelwise=st.booleans())
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_op.py:3822]: we can use this locally to make sure the implementation is good, but ideally we don't want to commit them to master, hypothesis tests has caused many issues in the past since it's not deterministic and always tries to find corner cases that the test fail, and typically it can find some since quantized kernels has rounding, clamping etc. which makes it not very numerically stable.

File: test/quantization/core/test_quantized_op.py [hunk 12]
```diff
@@ -3814,6 +3814,111 @@ def test_qlinear_unpack(self, W, use_channelwise):
...
+                                use_multi_dim_input, use_channelwise):
+        # Test numerical correctness of qlinear_leaky_relu
+        # Pass 1: X_q, W_q -> qlinear_leaky_relu -> Y_q
+        # Pass 2: X_q, W_q -> dequantize -> linear -> leaky_relu -> quantize -> Y_q_ref
+        # Compare the results of the two passes
+        decimal_val = 4
+        dtype = torch.quint8
+        negative_slope = 0.01
+        # for onednn backend only
+        with override_quantized_engine('onednn'):
```
Comment (jerryzh168) [inline test/quantization/core/test_quantized_op.py:3834]: are we still planning to use onednn, or do we want to switch to use x86 now?

File: torch/ao/quantization/backend_config/onednn.py [hunk 13]
```diff
@@ -0,0 +1,253 @@
+import torch
```
Comment (jerryzh168) [inline torch/ao/quantization/backend_config/onednn.py:1]: do we want to have a onednn backend? or just x86 backend?

**Comment by jerryzh168** [comment]
this PR is really large, I feel it might make sense to break this to
1). linear-bn-leakyrelu fusion op/kernel implementation and tests
2). defining backend config
3). test support in the quantization flow

File: test/quantization/core/test_quantized_op.py (hunk 12 above)
Comment (jgong5) [reply test/quantization/core/test_quantized_op.py:3834]: Thanks for point, @jerryzh168 . Ideally, we'd enable these new fusions on x86 directly. But the problem now is that the perf with onednn kernel cannot always be better than fbgemm even with post-op fusion. Therefore, our plan is to enable these new fusions on onednn backend first (which can benefit users in some cases but not all) while perf of onednn library being improved. Then, these new fusions can be moved to x86 when onednn kernel plus fusion can always bring better perf. Does this plan make sense to you?

**Comment by Xia-Weiwen** [comment]
> this PR is really large, I feel it might make sense to break this to 1). linear-bn-leakyrelu fusion op/kernel implementation and tests 2). defining backend config 3). test support in the quantization flow

Hi @jerryzh168 Thanks for your comments. As explained by Jiong and according to our previous discussion on Slack, we will continue improving onednn backend. I will close this PR later and split it into smaller ones.
